## 1.1 Basic Data Cleaning

In [33]:
import pandas as pd
import numpy as np

### PART 1: Removing unecessary columns <br>
- Browse data
- Find columns that are not useful/duplicate/same meaning as other columns
- Drop it

In [15]:
# Browse Data
df = pd.read_csv('BL-Flickr-Images-Book.csv')
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

__List all unuseful columns__<br>
For current dataset, if we, for example, just need to know description of each book, then columns Edition Statement, Corporate Author, Corporate Contributors, Former owner, Engraver, Issuance type and Shelfmarks are not relevant, remove them to reduce redundancy of dataset.

In [16]:
drop_list =['Edition Statement',
           'Corporate Author',
           'Corporate Contributors',
           'Former owner',
           'Engraver',
           'Contributors',
           'Issuance type',
           'Shelfmarks']
df.drop(drop_list, inplace=True, axis=1)
# inplace means drop on original dataset, axis=1 points to column name

Check if unuseful columns are removed successfully

In [17]:
list(df.columns) # all good

['Identifier',
 'Place of Publication',
 'Date of Publication',
 'Publisher',
 'Title',
 'Author',
 'Flickr URL']

### PART 2: Determine the index

- Set the column with unique value as the index and use it for query
- Not all dataset have such identifier
- One dataset may has more than one identifier

We can tell from column name to determine if it's a potential identifier <br>
Besides, loop can be applied to find identifier as well.

In [18]:
for column in df.columns:
    if df[column].is_unique:
        print("{} is eligible to be the identifier".format(column))

Identifier is eligible to be the identifier
Flickr URL is eligible to be the identifier


Since URL is weird to be the identifier, we choose column called Identifier to be identifier.

In [19]:
df.set_index('Identifier', inplace = True)

Check if we update dataset already

In [20]:
df.head() # all gooooooood

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

### PART 3: Data format clean up

__PART 3.1: Clean float data__<br>
From part 2, we use identifier(index) to get rows from table

In [21]:
# equivalent to sql:
# select Date_of_Publication from df where identifer > 1905 limit 5;
df.loc[1905:, 'Date of Publication'].head()

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
Name: Date of Publication, dtype: object

Format of 'Date of publication 'if not uniform, in this case, we only need its publication year - first 4 digit

In [24]:
df["Date of Publication"] = df["Date of Publication"].apply(lambda x: str(x[:4]))
# since data type of current column is float, we suppose to transfer it to string first, otherwise error will occur

In [25]:
df.loc[1905:, 'Date of Publication'].head()

Identifier
1905    1888
1929    1839
2836    1897
2854    1865
2956    1860
Name: Date of Publication, dtype: object

__PART 3.2: Clean text data__


In [30]:
df['Place of Publication'].head(30)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
1280                               Coventry
1808                            Christiania
1905                                Firenze
1929                              Amsterdam
2836                                 Savona
2854                                 London
2956                                  Paris
2957                                  Paris
3017                            Puerto-Rico
3131                               New York
4598                                   Hull
4884                                 London
4976                 

We need to uniform the address format: for address like "pp. 40. G. Bryan & Co: Oxford, 1898", we only remain the "oxford" part, for place who has a "-", we replace it as " ", we has 2 ways to realize this:

In [37]:
# Method one: np.where (equivilent to nested if else)
london = df["Place of Publication"].str.contains('London')
oxford = df['Place of Publication'].str.contains('Oxford')
df['Place of Publication'] = np.where(london, 'London',
                                    np.where(oxford, "Oxford", df["Place of Publication"].str.replace("-", " ")))
# check if the format has been cleaned up
df['Place of Publication'].head(10) # done :)

Identifier
206     London
216     London
218     London
472     London
480     London
481     London
519     London
667     Oxford
874     London
1143    London
Name: Place of Publication, dtype: object

In [40]:
# method two: apply() use if-else good for interprrtation
def transfer(x):
    x = str(x)
    if "London" in x:
        return "London"
    elif "Oxford" in x:
        return "Oxford"
    else:
        x.replace("-", " ")
        return x
df["Place of Publication"] = df["Place of Publication"].apply(lambda x: transfer(x))

### PART 4: Entire dataset cleanup

__university_towns is a txt file like below__<br>
Alabama[edit]<br>
Auburn (Auburn University)[1]<br>
Florence (University of North Alabama)<br>
Jacksonville (Jacksonville State University)[2]<br>
Livingston (University of West Alabama)[2]<br>
Montevallo (University of Montevallo)[2]<br>
Troy (Troy University)[2]<br>
...
In this case, we need to find the mapping from state to towns:
Alabama Auburn
Alabama Florence
....

In [47]:
map_town  = []
with  open("university_towns.txt") as file:
    for line in file:
        if '[edit]' in line:
            state = line[:line.find('[')]
        else:
            town = line[:line.find('(')]
        map_town.append([state, town])

In [48]:
df_town = pd.DataFrame(map_town, columns=['State', 'RegionName'])
df_town.head()

State     RegionName
0  Alabama       Laramie 
1  Alabama        Auburn 
2  Alabama      Florence 
3  Alabama  Jacksonville 
4  Alabama    Livingston